jni/src/faiss_index_service.cpp

```diff
@@ -57,76 +57,157 @@ void SetExtraParameters(knn_jni::JNIUtilInterface * jniUtil, JNIEnv *env,
 
     // Add vectors
-    std::unique_ptr<faiss::IndexIDMap> idMap(faissMethods->indexIdMap(indexWriter.get()));
     idMap->add_with_ids(numVectors, inputVectors->data(), ids.data());
+}
 
-    // Write the index to disk
-    faissMethods->writeIndex(idMap.get(), indexPath.c_str());
+void IndexService::writeIndex(
+        std::string indexPath,
+        jlong idMapAddress
+    ) {
+    std::unique_ptr<faiss::IndexIDMap> idMap (reinterpret_cast<faiss::IndexIDMap *> (idMapAddress));
+
+    try {
+        // Write the index to disk
+        faissMethods->writeIndex(idMap.get(), indexPath.c_str());
+    } catch(std::exception &e) {
+        delete idMap->index;
+        throw std::runtime_error("Failed to write index to disk");
+    }
+    // Free the memory used by the index
+    delete idMap->index;
 }
 
 BinaryIndexService::BinaryIndexService(std::unique_ptr<FaissMethods> faissMethods) : IndexService(std::move(faissMethods)) {}
 
-void BinaryIndexService::createIndex(
+void BinaryIndexService::allocIndex(faiss::Index * index, size_t dim, size_t numVectors) {
+    if(auto * indexBinaryHNSW = dynamic_cast<faiss::IndexBinaryHNSW *>(index)) {
+        auto * indexBinaryFlat = dynamic_cast<faiss::IndexBinaryFlat *>(indexBinaryHNSW->storage);
+        indexBinaryFlat->xb.reserve(dim * numVectors / 8);
+        return;
+    }
+}
+
+jlong BinaryIndexService::initIndex(
         knn_jni::JNIUtilInterface * jniUtil,
         JNIEnv * env,
         faiss::MetricType metric,
         std::string indexDescription,
         int dim,
-        int numIds,
+        int numVectors,
         int threadCount,
-        int64_t vectorsAddress,
-        std::vector<int64_t> ids,
-        std::string indexPath,
         std::unordered_map<std::string, jobject> parameters
     ) {
-    // Read vectors from memory address
-    auto *inputVectors = reinterpret_cast<std::vector<uint8_t>*>(vectorsAddress);
+    // Create index using Faiss factory method
+    std::unique_ptr<faiss::IndexBinary> indexWriter(faissMethods->indexBinaryFactory(dim, indexDescription.c_str()));
 
-    if (dim % 8 != 0) {
-        throw std::runtime_error("Dimensions should be multiply of 8");
+    // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+    if(threadCount != 0) {
+        omp_set_num_threads(threadCount);
     }
+
+    // Add extra parameters that cant be configured with the index factory
+    SetExtraParameters<faiss::IndexBinary, faiss::IndexBinaryIVF, faiss::IndexBinaryHNSW>(jniUtil, env, parameters, indexWriter.get());
+
+    // Check that the index does not need to be trained
+    if(!indexWriter->is_trained) {
+        throw std::runtime_error("Index is not trained");
+    }
+
+    std::unique_ptr<faiss::IndexBinaryIDMap> idMap(faissMethods->indexBinaryIdMap(indexWriter.get()));
+
+    allocIndex(dynamic_cast<faiss::Index *>(idMap->index), dim, numVectors);
+    indexWriter.release();
```

Comment:
do we need this?

Reply:
are you asking for why we need `allocIndex(dynamic_cast<faiss::Index *>(idMap->index), dim, numVectors);`?

Reply:
@jmazanec15 My understanding is `indexWriter.release();` is specifically done to not delete the the index created. The unique ptr will go out of scope once this method finishes, this is initializing the index, we need the index in future to make sure we can insert and write

Reply:
Added a comment to make it more evident